druid/src/menu/mod.rs

```diff
@@ -0,0 +1,762 @@
+//!                 // You need to handle the SHOW_PREFERENCES command yourself (or else do something
+//!                 // directly to the data here instead of using a command).
+//!                 .on_activate(|ctx, _data, _env| ctx.submit_command(commands::SHOW_PREFERENCES))
+//!                 .hotkey(SysMods::Cmd, ","),
+//!         )
+//!         .with_separator()
+//!         .with_entry(MenuItem::new(LocalizedString::new("macos-menu-services")))
+//!         .with_entry(
+//!             MenuItem::new(LocalizedString::new("macos-menu-hide-app"))
+//!                 // druid handles the HIDE_APPLICATION command automatically
+//!                 .on_activate(|ctx, _data, _env| ctx.submit_command(commands::HIDE_APPLICATION))
+//!                 .hotkey(SysMods::Cmd, "h"),
+//!         )
+//!         .with_entry(
+//!             MenuItem::new(LocalizedString::new("macos-menu-hide-others"))
+//!                 // druid handles the HIDE_OTHERS command automatically
+//!                 .on_activate(|ctx, _data, _env| ctx.submit_command(commands::HIDE_OTHERS))
+//!                 .hotkey(SysMods::AltCmd, "h"),
+//!         )
+//!         .with_entry(
+//!             MenuItem::new(LocalizedString::new("macos-menu-show-all"))
+//!                 // You need to handle the SHOW_ALL command yourself (or else do something
+//!                 // directly to the data here instead of using a command).
+//!                 .on_activate(|ctx, _data, _env| ctx.submit_command(commands::SHOW_ALL)),
+//!         )
+//!         .with_separator()
+//!         .with_entry(
+//!             MenuItem::new(LocalizedString::new("macos-menu-quit-app"))
+//!                 // druid handles the QUIT_APP command automatically
+//!                 .on_activate(|ctx, _data, _env| ctx.submit_command(commands::QUIT_APP))
+//!                 .hotkey(SysMods::Cmd, "q"),
+//!         )
+//! }
+//! ```
+//!
+//! [`LabelText::Dynamic`]: crate::widget::LabelText::Dynamic
+//! [`WindowDesc::menu`]: crate::WindowDesc::menu
+
+use std::num::NonZeroU32;
+
+use crate::core::CommandQueue;
+use crate::kurbo::Point;
+use crate::shell::{Counter, HotKey, IntoKey, Menu as PlatformMenu};
+use crate::widget::LabelText;
+use crate::{ArcStr, Command, Data, Env, Lens, RawMods};
+
+static COUNTER: Counter = Counter::new();
+
+pub mod sys;
+
+/// This is for completely recreating the menus (for when you want to change the actual menu
+/// structure, rather than just, say, enabling or disabling entries).
+pub(crate) struct MenuManager<T> {
+    build: Box<dyn FnMut(&T, &Env) -> Menu<T>>,
+    popup: bool,
+    old_data: Option<T>,
+    menu: Option<Menu<T>>,
+}
+
+/// A menu displayed as a pop-over.
+pub struct ContextMenu<T> {
+    pub(crate) build: Box<dyn FnMut(&T, &Env) -> Menu<T>>,
+    pub(crate) location: Point,
+}
+
+impl<T: Data> ContextMenu<T> {
+    /// Create a new [`ContextMenu`].
+    pub fn new(
```

Comment:
Just want to doublecheck my understanding here: I can construct my `ContextMenu` from *any* data, and not just the root app state, correct?

Reply:
No, currently the contexts' `show_context_menu` function still checks that we're using the root data. In principle, we could probably do something like what `new_sub_window` does, though.

Reply:
So playing around with this in runebender, I wonder if we can simplify the context menu case?

My basic thinking: a context menu is a modal. It should never need to be updated, since it is dismissed if anything is clicked. Additionally, in a non-trivial application a context menu is going to be presented somewhere fairly deep in the widget hierarchy, where the root data is likely to be several lenses away.

For me, I think that having access to the root data (especially when it's basically untyped) in the context menu just isn't that helpful. I would much prefer if, in this case, I instantiate a single menu object and display it each time.

Does this make sense? It's a bit different from the API for other menus, but ultimately feels simpler than what we have now and potentially more flexible, since you can pass in whatever data you want when building your menu.